qiita_pet/handlers/download.py

```diff
@@ -143,5 +143,87 @@ def get(self, extras):
+        # loop over artifacts and retrieve raw data (no parents)
+        to_download = []
+        vfabu = validate_filepath_access_by_user
+        for a in study.artifacts():
+            if not a.parents:
+                to_add = True
+                for i, (fid, path, data_type) in enumerate(a.filepaths):
+                    # validate access only of the first artifact filepath,
+                    # the rest have the same permissions
+                    if (i == 0 and not vfabu(user, fid)):
+                        to_add = False
+                        break
+                    if data_type == 'directory':
+                        # If we have a directory, we actually need to list
+                        # all the files from the directory so NGINX can
+                        # actually download all of them
+                        for dp, _, fps in walk(path):
+                            for fname in fps:
+                                fullpath = join(dp, fname)
+                                spath = fullpath
+                                if fullpath.startswith(basedir):
+                                    spath = fullpath[basedir_len:]
+                                to_download.append((fullpath, spath, spath))
+                    elif path.startswith(basedir):
+                        spath = path[basedir_len:]
+                        to_download.append((path, spath, spath))
+                    else:
+                        # We are not aware of any case that can trigger this
+                        # situation, but we wanted to be overly cautious
+                        # There is no test for this line cause we don't know
+                        # how to trigger it
+                        to_download.append((path, path, path))
 
+                if to_add:
+                    for pt in a.prep_templates:
+                        qmf = pt.qiime_map_fp
+                        if qmf is not None:
+                            sqmf = qmf
+                            if qmf.startswith(basedir):
+                                sqmf = qmf[basedir_len:]
+                            to_download.append(
+                                (qmf, sqmf, 'mapping_files/%s_mapping_file.txt'
+                                            % a.id))
+
+        # If we don't have nginx, write a file that indicates this
+        all_files = '\n'.join(["- %s /protected/%s %s" % (getsize(fp), sfp, n)
+                               for fp, sfp, n in to_download])
+        self.write("%s\n" % all_files)
+
+        zip_fn = 'study_raw_data_%d_%s.zip' % (
+            study_id, datetime.now().strftime('%m%d%y-%H%M%S'))
+
+        self.set_header('Content-Description', 'File Transfer')
+        self.set_header('Expires', '0')
+        self.set_header('Cache-Control', 'no-cache')
+        self.set_header('X-Archive-Files', 'zip')
+        self.set_header('Content-Disposition',
+                        'attachment; filename=%s' % zip_fn)
```

Comment:
Nowhere, is created on the fly by nginx. Basically, we send a list of filepaths to nginx using the "protected" filepath (only nginx has access) and the zip is created during download.